File: Codeforces/CF869C The Intriguing Obsession.py
max_n = 5005
mod = 998244353


fact = [1 for _ in xrange(max_n)]
for i in xrange(1, max_n):
    fact[i] = i * fact[i - 1]


def mod_rev(x):
    return pow(x, mod - 2, mod)


def mod_com(n, m):
    return fact[n] * mod_rev(fact[n - m] * fact[m]) % mod


# a, b, c = map(int, raw_input().split(' '))
a, b, c = 500, 500, 500

ans = 1
for x, y in [(a, b), (b, c), (c, a)]:
    num = 0
    for k in xrange(min(x, y) + 1):
        num += mod_com(x, k) * mod_com(y, k) * fact[k] % mod
    ans *= num
print ans % mod

ans = 1
for x, y in [(a, b), (b, c), (c, a)]:
    ans *= mod_com(max(x, y), min(x, y))
print ans % mod

File: Codeforces/CF869C The Intriguing Obsession.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAXN = 5000 + 5;
const int MOD = 998244353;

long long fact[MAXN];

long long modPow(long long x, long long y) {
    if (y == 0) {
        return 1;
    }
    long long ret = modPow(x * x % MOD, y >> 1);
    if (y & 1) {
        ret = ret * x % MOD;
    }
    return ret;
}

long long modRev(long long x) {
    return modPow(x, MOD - 2);
}

long long modCom(int n, int m) {
    return fact[n] * modRev(fact[n - m] * fact[m] % MOD) % MOD;
}

int main() {
    long long a, b, c;
    cin >> a >> b >> c;
    fact[0] = 1;
    for (int i = 1; i < MAXN; ++i) {
        fact[i] = fact[i - 1] * i % MOD;
    }
    long long ans = 1;
    for (auto p : vector<pair<int, int>>({{a, b}, {b, c}, {a, c}})) {
        long long num = 0;
        auto x = p.first;
        auto y = p.second;
        for (int k = 0; k <= min(x, y); ++k) {
            num += modCom(x, k) * modCom(y, k) % MOD * fact[k] % MOD;
        }
        ans = ans * (num % MOD) % MOD;
    }
    cout << ans << endl;
    return 0;
}
